homeassistant/components/zha/core/discovery.py

```diff
@@ -118,6 +195,125 @@ def discover_device_entities(self, device: ZHADevice) -> None:
 
         if device.is_coordinator:
             self.discover_coordinator_device_entities(device)
+            return
+
+        self.discover_quirks_v2_entities(device)
+        zha_regs.ZHA_ENTITIES.clean_up()
+
+    @callback
+    def discover_quirks_v2_entities(self, device: ZHADevice) -> None:
+        """Discover entities for a ZHA device exposed by quirks v2."""
+        _LOGGER.debug(
+            "Attempting to discover quirks v2 entities for device: %s-%s",
+            str(device.ieee),
+            device.name,
+        )
+
+        if not isinstance(device.device, CustomDeviceV2):
+            _LOGGER.debug(
+                "Device: %s-%s is not a quirks v2 device - skipping discover_quirks_v2_entities",
+                str(device.ieee),
+                device.name,
+            )
+            return
+
+        zigpy_device: CustomDeviceV2 = device.device
+
+        if not zigpy_device.exposes_metadata:
+            _LOGGER.debug(
+                "Device: %s-%s does not expose any quirks v2 entities",
+                str(device.ieee),
+                device.name,
+            )
+            return
+
+        for (
+            cluster_details,
+            quirk_metadata_list,
+        ) in zigpy_device.exposes_metadata.items():
+            endpoint_id, cluster_id, cluster_type = cluster_details
+
+            if endpoint_id not in device.endpoints:
+                _LOGGER.debug(
+                    "Device: %s-%s does not have an endpoint with id: %s - unable to create entity with cluster details: %s",
```

Comment:
Should this message be a warning maybe? Like, this shouldn't really happen with any zha-quirks release.
So, having a warning would make it clear why a custom entity isn't created for people developing a custom quirk.